### ISSUE TITLE
mutable_key_type error on tuple hashmap keys with generics

### ISSUE THREAD

**Issue by VictorKoenders** [rust-lang/rust-clippy#5043]
Clippy seems to give a false positive when a hashmap has a key that consists of a tuple of a struct and a generic:

```rust
struct SomeStruct<T: Hash + Eq> {
    map: HashMap<(T, U), ()>,
}

impl<T: Hash + Eq> SomeStruct<T> {
    fn test(&mut self) {
        let _mut_map_ref = &mut self.map;
        // error: mutable key type
        //   = note: `#[deny(clippy::mutable_key_type)]` on by default
        //   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#mutable_key_type
    }
}
```

Changing the key to be either `T` or `U` (not a tuple) removes the error:
```rust
// This is fine
struct SomeStruct<T: Hash + Eq> {
    map: HashMap<T, ()>,
}
// This is also fine
struct SomeStruct<T: Hash + Eq> {
    map: HashMap<U, ()>,
    pd: PhantomData<T>,
}
```

Playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=07ec1df6c1010ae39a5adf4017f0dca3

This might be related to #5020

**Comment by VictorKoenders** [comment]
Semi-related, in some cases the error can be mitigated by destructuring and then calling a mutable method on the hashmap reference.

![image](https://user-images.githubusercontent.com/2743142/72216994-cba87400-3528-11ea-9ff7-e76da50d5c28.png)
